tensornetwork/linalg/operations.py

```diff
@@ -0,0 +1,405 @@
+  errstr = ""
+  if not all_backends_same:
+    errstr = "All Tensors fed to " + fname + "must have the same backend."
+    errstr += "Backends were: \n"
+    errstr += str([name + "\n" for name in backend_names])
+  return all_backends_same, errstr
+
+
+def tensordot(a: Tensor, b: Tensor, axes: Sequence[Sequence[int]]) -> Tensor:
+  """Do a tensordot (contraction) of Tensors `a` and `b` over the given axes.
+  The behaviour of this function largely matches that of np.tensordot.
+
+  Args:
+    a: A Tensor.
+    b: Another Tensor.
+    axes: Two lists of integers. These values are the contraction
+          axes.
+  Raises:
+    ValueError, if a and b have different backends.
+  Returns:
+    The result of the tensordot, a Tensor.
+  """
+  if a.backend.name != b.backend.name:
+    errstr = "Tried to Tensordot Tensors with differing backends \n"
+    errstr += a.backend.name + "and " + b.backend.name + "."
+    raise ValueError(errstr)
+  out_array = a.backend.tensordot(a.array, b.array, axes)
+  out_tensor = Tensor(out_array, backend=a.backend)
+  return out_tensor
+
+
+def reshape(tensor: Tensor, new_shape: Sequence[int]) -> Tensor:
+  """Reshape Tensor to the given shape.
+
+  Args:
+    tensor: Tensor to reshape.
+    new_shape: The new shape.
+  Returns:
+    The reshaped Tensor.
+  """
+  return tensor.reshape(new_shape)
+
+
+def transpose(tensor: Tensor, perm: Optional[Sequence[int]] = None) -> Tensor:
+  """ Return a new `Tensor` transposed according to the permutation set
+  by `axes`. By default the axes are reversed.
+  Args:
+    axes: The permutation. If None (default) the index order is reversed.
+  Returns:
+    The transposed `Tensor`.
+  """
+  return tensor.transpose(perm=perm)
+
+
+def take_slice(tensor: Tensor, start_indices: Tuple[int, ...],
+               slice_sizes: Tuple[int, ...]) -> Tensor:
+  """Obtains a slice of a Tensor based on start_indices and slice_sizes.
+
+  Args:
+    Tensor: A Tensor.
+    start_indices: Tuple of integers denoting start indices of slice.
+    slice_sizes: Tuple of integers denoting size of slice along each axis.
+  Returns:
+    The slice, a Tensor.
+  """
+  sliced = tensor.backend.slice(tensor.array, start_indices, slice_sizes)
+  sliced_tensor = Tensor(sliced, backend=tensor.backend)
+  return sliced_tensor
+
+
+def shape(tensor: Tensor) -> Tuple[Optional[int], ...]:
+  """Get the shape of a Tensor as a tuple of integers.
+
+  Args:
+    Tensor: A Tensor.
+  Returns:
+    The shape of the input Tensor.
+  """
+  return tensor.shape
+
+
+def sqrt(tensor: Tensor) -> Tensor:
+  """Take the square root (element wise) of a given Tensor."""
+  out_array = tensor.backend.sqrt(tensor.array)
+  return Tensor(out_array, backend=tensor.backend)
+
+
+def outer(tensor1: Tensor, tensor2: Tensor) -> Tensor:
+  """Calculate the outer product of the two given Tensors."""
+  tensors = [tensor1, tensor2]
+  all_backends_same, errstr = _check_backends(tensors, "outer")
+  if not all_backends_same:
```

Comment:
is there a reason why you are not raising the error in `_check_backends`?

Reply:
Well, I think this would make the backtrace a bit harder to follow

Reply:
SG